File: ios/Beanshelf/Data/BeanPack.swift
import Foundation

/// Friend-to-friend bean sharing without a server: a `.beanshelf` file is one
/// JSON document with the bean's fields, brews, and both photos embedded as
/// base64. Send it over anything (AirDrop, Messages, email); the recipient's
/// Beanshelf imports it onto their own shelf. Same format as the Android app.
enum BeanPack {

    private static let version = 1

    static func export(_ bean: Bean) async -> URL? {
        await Task.detached(priority: .userInitiated) { () -> URL? in
            var root: [String: Any] = ["beanshelf": version]
            var b: [String: Any] = [
                "name": bean.name,
                "roaster": bean.roaster,
                "origin": bean.origin,
                "roastLevel": bean.roastLevel,
                "process": bean.process,
                "notes": bean.notes,
                "variety": bean.variety,
                "elevation": bean.elevation,
                "producer": bean.producer,
                "roastedOn": bean.roastedOn,
                "rating": bean.rating,
            ]
            b["brews"] = bean.brews.map { br -> [String: Any] in
                var o: [String: Any] = [
                    "method": br.method,
                    "rating": br.rating,
                    "note": br.note,
                    "timestamp": br.timestamp,
                    "grinder": br.grinder,
                    "grindSize": br.grindSize,
                ]
                o["doseG"] = br.doseG ?? NSNull()
                o["waterG"] = br.waterG ?? NSNull()
                return o
            }
            root["bean"] = b
            if let file = bean.photoFile, let url = PhotoStore.url(for: file),
               let data = try? Data(contentsOf: url) {
                root["frontPhoto"] = data.base64EncodedString()
                root["frontIsCutout"] = BagCropper.isCutout(file)
            }
            if let file = bean.backPhotoFile, let url = PhotoStore.url(for: file),
               let data = try? Data(contentsOf: url) {
                root["backPhoto"] = data.base64EncodedString()
                root["backIsCutout"] = BagCropper.isCutout(file)
            }

            guard let json = try? JSONSerialization.data(withJSONObject: root) else { return nil }
            var safeName = bean.name.isEmpty ? "bean" : bean.name
            safeName = safeName.replacingOccurrences(of: #"[^A-Za-z0-9 _-]"#, with: "", options: .regularExpression)
                .trimmingCharacters(in: .whitespaces)
                .replacingOccurrences(of: " ", with: "-")
                .lowercased()
            if safeName.isEmpty { safeName = "bean" }
            let dir = FileManager.default.temporaryDirectory.appendingPathComponent("share", isDirectory: true)
            try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
            let out = dir.appendingPathComponent("\(safeName).beanshelf")
            do {
                try json.write(to: out, options: .atomic)
                return out
            } catch {
                return nil
            }
        }.value
    }

    /// Reads a .beanshelf file into a NEW bean (fresh id, photos copied in).
    static func `import`(url: URL) async -> Bean? {
        await Task.detached(priority: .userInitiated) { () -> Bean? in
            let secured = url.startAccessingSecurityScopedResource()
            defer { if secured { url.stopAccessingSecurityScopedResource() } }
            guard let data = try? Data(contentsOf: url),
                  let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  (root["beanshelf"] as? Int ?? 0) >= 1,
                  let b = root["bean"] as? [String: Any]
            else { return nil }

            func savePhoto(_ key: String, _ cutoutKey: String) -> String? {
                guard let b64 = root[key] as? String, !b64.isEmpty,
                      let bytes = Data(base64Encoded: b64) else { return nil }
                let ext = (root[cutoutKey] as? Bool ?? false) ? "png" : "jpg"
                let name = "\(UUID().uuidString).\(ext)"
                guard let dest = PhotoStore.url(for: name),
                      (try? bytes.write(to: dest, options: .atomic)) != nil else { return nil }
                return name
            }

            let brews = (b["brews"] as? [[String: Any]] ?? []).map { br in
                Brew(
                    id: UUID().uuidString,
                    method: br["method"] as? String ?? "",
                    rating: br["rating"] as? Double ?? 0,
                    note: br["note"] as? String ?? "",
                    timestamp: (br["timestamp"] as? NSNumber)?.int64Value ?? 0,
                    doseG: br["doseG"] as? Double,
                    waterG: br["waterG"] as? Double,
                    grinder: br["grinder"] as? String ?? "",
                    grindSize: br["grindSize"] as? String ?? ""
                )
            }
            return Bean(
                id: UUID().uuidString,
                name: b["name"] as? String ?? "",
                roaster: b["roaster"] as? String ?? "",
                origin: b["origin"] as? String ?? "",
                roastLevel: b["roastLevel"] as? String ?? "",
                process: b["process"] as? String ?? "",
                notes: b["notes"] as? String ?? "",
                variety: b["variety"] as? String ?? "",
                elevation: b["elevation"] as? String ?? "",
                producer: b["producer"] as? String ?? "",
                roastedOn: b["roastedOn"] as? String ?? "",
                rating: b["rating"] as? Double ?? 0,
                photoFile: savePhoto("frontPhoto", "frontIsCutout"),
                backPhotoFile: savePhoto("backPhoto", "backIsCutout"),
                createdAt: nowMillis(),
                brews: brews
            )
        }.value
    }
}

File: app/src/main/java/dev/adamsjack/beanshelf/data/BeanPack.kt
package dev.adamsjack.beanshelf.data

import android.content.Context
import android.net.Uri
import android.util.Base64
import dev.adamsjack.beanshelf.model.Bean
import dev.adamsjack.beanshelf.model.Brew
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.util.UUID

/**
 * Friend-to-friend bean sharing without a server: a `.beanshelf` file is one
 * JSON document with the bean's fields, brews, and both photos embedded as
 * base64. Send it over anything (AirDrop, chat, email); the recipient's
 * Beanshelf imports it onto their own shelf.
 */
object BeanPack {

    private const val VERSION = 1

    suspend fun export(context: Context, bean: Bean): File = withContext(Dispatchers.IO) {
        val root = JSONObject().apply {
            put("beanshelf", VERSION)
            put("bean", JSONObject().apply {
                put("name", bean.name)
                put("roaster", bean.roaster)
                put("origin", bean.origin)
                put("roastLevel", bean.roastLevel)
                put("process", bean.process)
                put("notes", bean.notes)
                put("variety", bean.variety)
                put("elevation", bean.elevation)
                put("producer", bean.producer)
                put("roastedOn", bean.roastedOn)
                put("rating", bean.rating.toDouble())
                put("brews", JSONArray().apply {
                    bean.brews.forEach { br ->
                        put(JSONObject().apply {
                            put("method", br.method)
                            put("rating", br.rating.toDouble())
                            put("note", br.note)
                            put("timestamp", br.timestamp)
                            put("doseG", br.doseG?.toDouble() ?: JSONObject.NULL)
                            put("waterG", br.waterG?.toDouble() ?: JSONObject.NULL)
                            put("grinder", br.grinder)
                            put("grindSize", br.grindSize)
                        })
                    }
                })
            })
            bean.photoPath?.let { p ->
                put("frontPhoto", Base64.encodeToString(File(p).readBytes(), Base64.NO_WRAP))
                put("frontIsCutout", BagCropper.isCutout(p))
            }
            bean.backPhotoPath?.let { p ->
                put("backPhoto", Base64.encodeToString(File(p).readBytes(), Base64.NO_WRAP))
                put("backIsCutout", BagCropper.isCutout(p))
            }
        }
        val dir = File(context.cacheDir, "share").apply { mkdirs() }
        val safeName = bean.name.ifBlank { "bean" }.replace(Regex("[^A-Za-z0-9 _-]"), "").trim()
            .replace(' ', '-').lowercase().ifBlank { "bean" }
        val out = File(dir, "$safeName.beanshelf")
        out.writeText(root.toString())
        out
    }

    /** Reads a .beanshelf file into a NEW bean (fresh id, photos copied in). */
    suspend fun import(context: Context, uri: Uri): Bean? = withContext(Dispatchers.IO) {
        runCatching {
            val text = context.contentResolver.openInputStream(uri)?.use {
                it.readBytes().toString(Charsets.UTF_8)
            } ?: return@runCatching null
            val root = JSONObject(text)
            if (root.optInt("beanshelf", 0) < 1) return@runCatching null
            val b = root.getJSONObject("bean")

            fun savePhoto(key: String, cutoutKey: String): String? {
                val b64 = root.optString(key, "")
                if (b64.isBlank()) return null
                val bytes = Base64.decode(b64, Base64.NO_WRAP)
                val ext = if (root.optBoolean(cutoutKey, false)) "png" else "jpg"
                val f = File(PhotoStore.photosDir(context), "${UUID.randomUUID()}.$ext")
                f.writeBytes(bytes)
                return f.absolutePath
            }

            val brewsArr = b.optJSONArray("brews") ?: JSONArray()
            Bean(
                id = UUID.randomUUID().toString(),
                name = b.optString("name"),
                roaster = b.optString("roaster"),
                origin = b.optString("origin"),
                roastLevel = b.optString("roastLevel"),
                process = b.optString("process"),
                notes = b.optString("notes"),
                variety = b.optString("variety"),
                elevation = b.optString("elevation"),
                producer = b.optString("producer"),
                roastedOn = b.optString("roastedOn"),
                rating = b.optDouble("rating", 0.0).toFloat(),
                photoPath = savePhoto("frontPhoto", "frontIsCutout"),
                backPhotoPath = savePhoto("backPhoto", "backIsCutout"),
                createdAt = System.currentTimeMillis(),
                brews = (0 until brewsArr.length()).map { i ->
                    val br = brewsArr.getJSONObject(i)
                    Brew(
                        id = UUID.randomUUID().toString(),
                        method = br.optString("method"),
                        rating = br.optDouble("rating", 0.0).toFloat(),
                        note = br.optString("note"),
                        timestamp = br.optLong("timestamp"),
                        doseG = if (br.isNull("doseG")) null else br.optDouble("doseG").toFloat(),
                        waterG = if (br.isNull("waterG")) null else br.optDouble("waterG").toFloat(),
                        grinder = br.optString("grinder"),
                        grindSize = br.optString("grindSize"),
                    )
                },
            )
        }.getOrNull()
    }
}
